### ISSUE TITLE
Setup a database with OSM-derived data.

### ISSUE THREAD

**Issue by xivk** [M1dgard/OsmNewContributors#1]
The first thing we need to do is setup a database that contains all basic data about nodes locations, ways and relations.

As we review a changeset there is a need to geographically locate the change. When using a default planet-database there are three options:

- A change to a node: Easy, just use the lat/lon in the changeset to determine location.
- A change to a way: Enumerate all nodes and get their location from a database containing all nodes. Then use their lat/lon to determine location.
- A change to a relation: Enumerate member and get their location from a database containing all nodes/ways/relations.

There is also the option to setup a database where we store only a node/way/relation's boundingbox and we could keep this up-to-date by using the minutely changes.

**Comment by M1dgard** [comment]
Do we need to keep a spatial database? Changesets themselves are replicated in http://planet.openstreetmap.org/replication/changesets/001/, including the bounding box. Does storing them in a geo-database have additional benefits? Or is this for data validation?

Django has a built-in GIS framework: [GeoDjango](https://docs.djangoproject.com/en/1.8/ref/contrib/gis/).

**Comment by xivk** [comment]
If i'm not mistaken these boxes contain the boundingbox for the entire changeset. Meaning, If I make a change in the north of france and one in germany and then upload to the OSM-API the box will look like this:

http://geojson.io/#id=gist:anonymous/9999bd6048a6928f34ed&map=6/50.212/1.846

This without even one edit in Belgium. This happens a lot, have a look at the history tab on the OSM-website. Example:

http://www.openstreetmap.org/changeset/33411761

**Comment by M1dgard** [comment]
Ah, yes, I did not think of that to happen on a first edit but it may. Shall we initially populate the database with a planet dump or only track new edits from the minutely? Am I correct in thinking that we don't need a planet, because the minutely diffs contain full geographic information about all relevant objects?

**Comment by xivk** [comment]
Changesets contain only the modified object. This means when a way is modified but adding a new tag then only the way is in the changeset, not all the nodes. This means we cannot know where the way is located in the world without a full dump of the database.

As i see it there are two options:
- Use a standard planet dump (limited to Belgium for now).
- Use a customized database only keeping boundingboxes per node/way/relation.

Option 1 has the advantage that we have access to the complete database to do vandalism detect but It will require a very big database.
Option 2 has the advantage that it's (a lot) smaller but it requires some custom development.

**Comment by M1dgard** [comment]
> when a way is modified by adding a new tag then only the way is in the changeset, not all the nodes.

That's what I feared.

In option 2, how do we know what the new bounding box of a way is, if we only have a list of the nodes' IDs? If I'm not mistaken, a member list with all children's IDs is included every time for ways and relations. As you say, their geometry isn't.

To illustrate this, consider a simple way W with 3 nodes, named A, B, C and D for convenience.

```txt
  A               B
          D
  C
```

Joe creates the way. Fine, you can easily calculate the bounding box. But now Patrick moves C:

```txt
  A       C       B
          D
```

We have no way to know what the bounding box looks like, since we haven't remembered the position of D.

I think we can only choose between naive changeset bounding boxes and full database replication. (Initially maybe only for Belgium.)

Only looking at changed nodes is of course no viable option, you can change the tags of 100 ways and just as many relations without ever touching a node.

**Comment by M1dgard** [comment]
Ah, no, we *do* have D's coordinates, and the full list of nodes is includes with every change so we don't have to remember ... Now I see it would be possible.

**Comment by xivk** [comment]
Hmm.... you made me realize, we do still need to keep a list of all nodes around per way and a list of members per relations. If a way changes as a member of a relation or a node changes as a member of a relation and it changes the bb of the parent we need to be able to update it's parent too.

**Comment by M1dgard** [comment]
I got confused halfway writing my explanation, and forgot the reason why it is not possible. :p

I've corrected my comments for future reference.

Then the only option is to do full replication, right?

**Comment by xivk** [comment]
Or some form of replication with keeping everything around. But the replication tools are already there so why not just use them for now?

I guess it all depends on what we want to build: a dedicated tool for welcome messages or do we also want to do other stuff with this in the future and ifso what?

**Comment by M1dgard** [comment]
Just an (very-)long-term idea: we could build in a form of automatic recognition of potentially problematic edits. Or try to automatically classify what a mapper has done. (Map houses, change POI details ...) If OpenStreetMap would receive more and more new mappers, this could be a real help for the welcomers. But if this is ever going to happen, it's a long time from now.

For now, a system to simply manage the messages and keep an eye on the newbies will suffice, as outlined in [Primary goals](https://wiki.openstreetmap.org/wiki/Collaborative_tool_for_welcoming_new_mappers#Primary_goals), while also keeping the secondary goal and the other ideas in mind.

**Comment by M1dgard** [comment]
Can we use [osm2pgsql](https://wiki.openstreetmap.org/wiki/Osm2pgsql) to import the data?

**Comment by ethan-nelson** [comment]
With my work in the DWG, I have been building a tool to detect X or Y types of edits. It's very rough around the edges and definitely in active development as I type (between dissertation and work), but you can check it out at https://github.com/ethan-nelson/osm_hall_monitor. My goal is to build the tool to be generalized enough such that it could be applied to something like this. The way I see it:

1. Scan hourly diffs 
2. Look at the users for each changeset and compile their information (join date, number of previous edits, etc.)
3. If they are new`*`, send a notification to a person or register it in the db and show up on a Django page.
4. Repeat

`*` - We can look at the number of changesets submitted in the hour through the diff and compare it to the number of edits after polling the user api. We would want to include some wiggle room, like maybe 5 changesets, so that if they submitted more edits between the time the diff was written and the time we download that file, process it, and check their user info they wouldn't slip through the cracks. Sure, this may create some false positives, but those can be easily sifted through unless we don't mind the occasional user falling through the cracks.

Yes, changesets' spans are quite vast in size, but I would argue new users probably are making a single edit or two in their first changesets. So, we should be able to use the changeset bbox or information on the nodes to somewhat ascertain their location (assuming they are editing generally in the same region they live). If not, we can further process the changesets (again, another aspect that will hopefully be built into my tool soon) to try and find the area where the majority of edits are made.

**Comment by M1dgard** [comment]
> I'm not necessarily sure we would need to use up the space to create an identical OSM database--unless that's something you're fine with doing.

Indeed, I've been reading up on OSM databases and a full database replication is not really doable.
  
> We can look at the number of changesets submitted in the hour through the diff and compare it to the number of edits after polling the user api. We would want to include some wiggle room, like maybe 5 changesets, so that if they submitted more edits between the time the diff was written and the time we download that file, process it, and check their user info they wouldn't slip through the cracks. Sure, this may create some false positives, but those can be easily sifted through unless we don't mind the occasional user falling through the cracks.

The user API is kind of slow if you have to query a lot of users. What I had in mind was making a list of known users and marking them as new if
a) their user ID > a fixed one (the "user ID of the day the program starts") and
b) they aren't on the list.

> Yes, changesets' spans are quite vast in size, but I would argue new users probably are making a single edit or two in their first changesets.

That's also what I was thinking at first. :)

> If not, we can further process the changesets (again, another aspect that will hopefully be built into my tool soon) to try and find the area where the majority of edits are made.

Do you have an idea on how this could be achieved without keeping a copy of each OSM object?

Thanks for your interest and suggestions!

**Comment by M1dgard** [comment]
> Can we use osm2pgsql to import the data?

It's more to prepare a render DB and does not support updating. Osmosis is better for that.
I found https://wiki.openstreetmap.org/wiki/Databases_and_data_access_APIs#Database_Schemas but I'm not sure what would be suitable.

**Comment by ethan-nelson** [comment]
@M1dgard I will check more this weekend what all I can extract from the diff files and maybe run a few tests with my workflow about timing, etc. In theory, I do like your idea about the user id being greater than a known one for a given day, but one thing that may convolute that is the archetypal user that registered a year ago but just started editing today.

**Comment by M1dgard** [comment]
@ethan-nelson True. In our manual welcoming list we have quite a lot of those.
